CS4001/4042 Assignment 1, Part B, Q1
---

Real world datasets often have a mix of numeric and categorical features – this dataset is one example. To build models on such data, categorical features have to be encoded or embedded.

PyTorch Tabular is a library that makes it very convenient to build neural networks for tabular data. It is built on top of PyTorch Lightning, which abstracts away boilerplate model training code and makes it easy to integrate other tools, e.g. TensorBoard for experiment tracking.

For questions B1 and B2, the following features should be used:   
- **Numeric / Continuous** features: dist_to_nearest_stn, dist_to_dhoby, degree_centrality, eigenvector_centrality, remaining_lease_years, floor_area_sqm
- **Categorical** features: month, town, flat_model_type, storey_range



---



In [1]:
!pip install pytorch_tabular[extra]

  Using cached torch-1.13.1-cp310-cp310-win_amd64.whl (162.6 MB)
  Attempting uninstall: torch
    Found existing installation: torch 2.0.1
    Uninstalling torch-2.0.1:
      Successfully uninstalled torch-2.0.1


In [2]:
SEED = 42

import os

import random
random.seed(SEED)

import numpy as np
np.random.seed(SEED)

import pandas as pd

from pytorch_tabular import TabularModel
from pytorch_tabular.models import CategoryEmbeddingModelConfig
from pytorch_tabular.config import (
    DataConfig,
    OptimizerConfig,
    TrainerConfig,
)

> Divide the dataset (‘hdb_price_prediction.csv’) into train, validation and test sets by using entries from year 2019 and before as training data, year 2020 as validation data and year 2021 as test data.
**Do not** use data from year 2022 and year 2023.



In [3]:
df = pd.read_csv('hdb_price_prediction.csv')

train_df = df[df['year'] <= 2019]
val_df = df[df['year'] == 2020]
test_df = df[df['year'] == 2021]

train_df

month  year        town              full_address   nearest_stn  \
0          1  2017  ANG MO KIO  406 ANG MO KIO AVENUE 10    Ang Mo Kio   
1          1  2017  ANG MO KIO   108 ANG MO KIO AVENUE 4    Ang Mo Kio   
2          1  2017  ANG MO KIO   602 ANG MO KIO AVENUE 5  Yio Chu Kang   
3          1  2017  ANG MO KIO  465 ANG MO KIO AVENUE 10    Ang Mo Kio   
4          1  2017  ANG MO KIO   601 ANG MO KIO AVENUE 5  Yio Chu Kang   
...      ...   ...         ...                       ...           ...   
64052     12  2019      YISHUN      326 YISHUN RING ROAD        Yishun   
64053     12  2019      YISHUN      326 YISHUN RING ROAD        Yishun   
64054     12  2019      YISHUN      608 YISHUN STREET 61        Khatib   
64055     12  2019      YISHUN      604 YISHUN STREET 61        Khatib   
64056     12  2019      YISHUN      877 YISHUN STREET 81        Khatib   

       dist_to_nearest_stn  dist_to_dhoby  degree_centrality  \
0                 1.007264       7.006044           0.016807   
1                 1.271389       7.983837           0.016807   
2                 1.069743       9.090700           0.016807   
3                 0.946890       7.519889           0.016807   
4                 1.092551       9.130489           0.016807   
...                    ...            ...                ...   
64052             0.823163      14.421823           0.016807   
64053             0.823163      14.421823           0.016807   
64054             0.445869      13.498243           0.016807   
64055             0.552769      13.598257           0.016807   
64056             0.475885      12.738721           0.016807   

       eigenvector_centrality         flat_model_type  remaining_lease_years  \
0                    0.006243        2 ROOM, Improved              61.333333   
1                    0.006243  3 ROOM, New Generation              60.583333   
2                    0.002459  3 ROOM, New Generation              62.416667   
3                    0.006243  3 ROOM, New Generation              62.083333   
4                    0.002459  3 ROOM, New Generation              62.416667   
...                       ...                     ...                    ...   
64052                0.000382    EXECUTIVE, Apartment              67.583333   
64053                0.000382   EXECUTIVE, Maisonette              67.583333   
64054                0.000968    EXECUTIVE, Apartment              71.500000   
64055                0.000968    EXECUTIVE, Apartment              71.500000   
64056                0.000968   EXECUTIVE, Maisonette              67.083333   

       floor_area_sqm storey_range  resale_price  
0                44.0     10 TO 12      232000.0  
1                67.0     01 TO 03      250000.0  
2                67.0     01 TO 03      262000.0  
3                68.0     04 TO 06      265000.0  
4                67.0     01 TO 03      265000.0  
...               ...          ...           ...  
64052           142.0     04 TO 06      580000.0  
64053           146.0     07 TO 09      565000.0  
64054           164.0     01 TO 03      633000.0  
64055           164.0     10 TO 12      788888.0  
64056           145.0     07 TO 09      618000.0  

[64057 rows x 14 columns]

> Refer to the documentation of **PyTorch Tabular** and perform the following tasks: https://pytorch-tabular.readthedocs.io/en/latest/#usage
- Use **[DataConfig](https://pytorch-tabular.readthedocs.io/en/latest/data/)** to define the target variable, as well as the names of the continuous and categorical variables.
- Use **[TrainerConfig](https://pytorch-tabular.readthedocs.io/en/latest/training/)** to automatically tune the learning rate. Set batch_size to be 1024 and set max_epoch as 50.
- Use **[CategoryEmbeddingModelConfig](https://pytorch-tabular.readthedocs.io/en/latest/models/#category-embedding-model)** to create a feedforward neural network with 1 hidden layer containing 50 neurons.
- Use **[OptimizerConfig](https://pytorch-tabular.readthedocs.io/en/latest/optimizer/)** to choose Adam optimiser. There is no need to set the learning rate (since it will be tuned automatically) nor scheduler.
- Use **[TabularModel](https://pytorch-tabular.readthedocs.io/en/latest/tabular_model/)** to initialise the model and put all the configs together.

In [4]:
# TODO: Enter your code here
num_col_names = ["dist_to_nearest_stn", "dist_to_dhoby", "degree_centrality", "eigenvector_centrality", "remaining_lease_years", "floor_area_sqm"]
cat_col_names = ["month", "town", "flat_model_type", "storey_range"]

data_config = DataConfig(
    target=[
        "resale_price"
    ],  # target should always be a list. Multi-targets are only supported for regression. Multi-Task Classification is not implemented
    continuous_cols=num_col_names,
    categorical_cols=cat_col_names,
)

trainer_config = TrainerConfig(
    auto_lr_find=True,  # Runs the LRFinder to automatically derive a learning rate
    batch_size=1024,
    max_epochs=50,
)

model_config = CategoryEmbeddingModelConfig(
    task="regression",
    layers="50",  # Number of nodes in each layer
    activation="ReLU"  # Activation between each layers
)

optimizer_config = OptimizerConfig(optimizer='Adam')

tabular_model = TabularModel(
    data_config=data_config,
    model_config=model_config,
    optimizer_config=optimizer_config,
    trainer_config=trainer_config,
)

tabular_model.fit(train=train_df, validation=val_df)
result = tabular_model.evaluate(test_df)
pred_df = tabular_model.predict(test_df)
rmse = np.sqrt(result[0]['test_mean_squared_error'])
def r_squared_numpy(y_true, y_pred):
    y_mean = np.mean(y_true)
    ss_tot = np.sum((y_true - y_mean)**2)
    ss_res = np.sum((y_true - y_pred)**2)
    r2 = 1 - (ss_res / ss_tot)
    return r2
ground_truth = pred_df['resale_price'].to_numpy()
pred = pred_df['resale_price_prediction'].to_numpy()

r2 = r_squared_numpy(ground_truth, pred)
print("RMSE: ", rmse)
print("R2: ", r2)

2023-10-03 16:21:31,443 - {pytorch_tabular.tabular_model:105} - INFO - Experiment Tracking is turned off
Global seed set to 42
2023-10-03 16:21:31,468 - {pytorch_tabular.tabular_model:473} - INFO - Preparing the DataLoaders
2023-10-03 16:21:31,470 - {pytorch_tabular.tabular_datamodule:290} - INFO - Setting up the datamodule for regression task
2023-10-03 16:21:31,560 - {pytorch_tabular.tabular_model:521} - INFO - Preparing the Model: CategoryEmbeddingModel
2023-10-03 16:21:31,585 - {pytorch_tabular.tabular_model:268} - INFO - Preparing the Trainer
c:\Users\mdzak\anaconda3\envs\nnpartb\lib\site-packages\pytorch_lightning\trainer\connectors\accelerator_connector.py:589: LightningDeprecationWarning: The Trainer argument `auto_select_gpus` has been deprecated in v1.9.0 and will be removed in v2.0.0. Please use the function `pytorch_lightning.accelerators.find_usable_cuda_devices` instead.
  rank_zero_deprecation(
GPU available: False, used: False
TPU available: False, using: 0 TPU cores
IP

Finding best initial lr:   0%|          | 0/100 [00:00<?, ?it/s]

c:\Users\mdzak\anaconda3\envs\nnpartb\lib\site-packages\pytorch_lightning\core\module.py:493: UserWarning: You called `self.log('train_loss', ..., logger=True)` but have no logger configured. You can enable one by doing `Trainer(logger=ALogger(...))`
  rank_zero_warn(
c:\Users\mdzak\anaconda3\envs\nnpartb\lib\site-packages\pytorch_lightning\core\module.py:493: UserWarning: You called `self.log('train_mean_squared_error', ..., logger=True)` but have no logger configured. You can enable one by doing `Trainer(logger=ALogger(...))`
  rank_zero_warn(
`Trainer.fit` stopped: `max_steps=100` reached.
Learning rate set to 0.5754399373371567
Restoring states from the checkpoint path at c:\Users\mdzak\Desktop\GitHub\CZ4042_NN\submission\.lr_find_c31433a4-199d-43fc-b76a-ef16bd24c3eb.ckpt
Restored all states from the checkpoint file at c:\Users\mdzak\Desktop\GitHub\CZ4042_NN\submission\.lr_find_c31433a4-199d-43fc-b76a-ef16bd24c3eb.ckpt
2023-10-03 16:21:34,667 - {pytorch_tabular.tabular_model:575} -

┏━━━┳━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━┓
┃   ┃ Name             ┃ Type                      ┃ Params ┃
┡━━━╇━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━┩
│ 0 │ _backbone        │ CategoryEmbeddingBackbone │  2.9 K │
│ 1 │ _embedding_layer │ Embedding1dLayer          │  1.5 K │
│ 2 │ head             │ LinearHead                │     51 │
│ 3 │ loss             │ MSELoss                   │      0 │
└───┴──────────────────┴───────────────────────────┴────────┘

Trainable params: 4.5 K                                                                                            
Non-trainable params: 0                                                                                            
Total params: 4.5 K                                                                                                
Total estimated model params size (MB): 0

Output()

2023-10-03 16:21:54,037 - {pytorch_tabular.tabular_model:584} - INFO - Training the model completed
2023-10-03 16:21:54,039 - {pytorch_tabular.tabular_model:1258} - INFO - Loading the best model


Output()

c:\Users\mdzak\anaconda3\envs\nnpartb\lib\site-packages\pytorch_lightning\core\module.py:493: UserWarning: You 
called `self.log('test_loss', ..., logger=True)` but have no logger configured. You can enable one by doing 
`Trainer(logger=ALogger(...))`
  rank_zero_warn(

c:\Users\mdzak\anaconda3\envs\nnpartb\lib\site-packages\pytorch_lightning\core\module.py:493: UserWarning: You 
called `self.log('test_mean_squared_error', ..., logger=True)` but have no logger configured. You can enable one by
doing `Trainer(logger=ALogger(...))`
  rank_zero_warn(

c:\Users\mdzak\anaconda3\envs\nnpartb\lib\site-packages\pytorch_lightning\utilities\cloud_io.py:33: LightningDeprecationWarning: `pytorch_lightning.utilities.cloud_io.get_filesystem` has been deprecated in v1.8.0 and will be removed in v2.0.0. Please use `lightning_fabric.utilities.cloud_io.get_filesystem` instead.
  rank_zero_deprecation(
c:\Users\mdzak\anaconda3\envs\nnpartb\lib\site-packages\pytorch_lightning\trainer\connectors\data_connector.py:224: PossibleUserWarning: The dataloader, test_dataloader 0, does not have many workers which may be a bottleneck. Consider increasing the value of the `num_workers` argument` (try 16 which is the number of cpus on this machine) in the `DataLoader` init to improve performance.
  rank_zero_warn(


┏━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
┃        Test metric        ┃       DataLoader 0        ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━┩
│         test_loss         │       5882418688.0        │
│  test_mean_squared_error  │       5882418688.0        │
└───────────────────────────┴───────────────────────────┘

Output()

RMSE:  76696.92750038947
R2:  0.7776187406639129


> Report the test RMSE error and the test R2 value that you obtained.



<!-- \# TODO: \<Enter your answer here\> -->

RMSE: 76696.92750038947  
R2: 0.7776187406639129


> Print out the corresponding rows in the dataframe for the top 25 test samples with the largest errors. Identify a trend in these poor predictions and suggest a way to reduce these errors.



In [9]:
pred_df_25 = pred_df
pred_df_25['abs_sq_error'] = (pred_df_25['resale_price'] - pred_df_25['resale_price_prediction'])**2
sorted_pred_df_25 = pred_df_25.sort_values(by='abs_sq_error', ascending=False)

sorted_pred_df_25.head(25)


month  year          town                full_address    nearest_stn  \
92405      11  2021   BUKIT MERAH            46 SENG POH ROAD    Tiong Bahru   
90957       6  2021   BUKIT BATOK  288A BUKIT BATOK STREET 25    Bukit Batok   
112128     12  2021      TAMPINES      156 TAMPINES STREET 12       Tampines   
90608      12  2021        BISHAN       273B BISHAN STREET 24         Bishan   
106192     12  2021    QUEENSTOWN              89 DAWSON ROAD     Queenstown   
91871       6  2021   BUKIT MERAH         17 TIONG BAHRU ROAD    Tiong Bahru   
93825       8  2021  CENTRAL AREA       4 TANJONG PAGAR PLAZA  Tanjong Pagar   
92504      12  2021   BUKIT MERAH            49 KIM PONG ROAD    Tiong Bahru   
105695      6  2021    QUEENSTOWN              91 DAWSON ROAD     Queenstown   
90432       8  2021        BISHAN       275A BISHAN STREET 24         Bishan   
92299      10  2021   BUKIT MERAH         36 MOH GUAN TERRACE    Tiong Bahru   
92442      11  2021   BUKIT MERAH          127D KIM TIAN ROAD    Tiong Bahru   
93895      11  2021  CENTRAL AREA       3 TANJONG PAGAR PLAZA  Tanjong Pagar   
105699      6  2021    QUEENSTOWN              91 DAWSON ROAD     Queenstown   
88081       8  2021    ANG MO KIO    310A ANG MO KIO AVENUE 1     Ang Mo Kio   
106057     10  2021    QUEENSTOWN              92 DAWSON ROAD     Queenstown   
90521      10  2021        BISHAN        237 BISHAN STREET 22         Bishan   
92340      10  2021   BUKIT MERAH            56 HAVELOCK ROAD    Tiong Bahru   
91694       4  2021   BUKIT MERAH          35 LIM LIAK STREET    Tiong Bahru   
98379      12  2021       HOUGANG        615 HOUGANG AVENUE 8        Hougang   
92226       9  2021   BUKIT MERAH          96A HENDERSON ROAD    Tiong Bahru   
105468      3  2021    QUEENSTOWN              90 DAWSON ROAD     Queenstown   
92533      12  2021   BUKIT MERAH          2C BOON TIONG ROAD    Tiong Bahru   
105966      9  2021    QUEENSTOWN              91 DAWSON ROAD     Queenstown   
92066       8  2021   BUKIT MERAH         48 MOH GUAN TERRACE    Tiong Bahru   

        dist_to_nearest_stn  dist_to_dhoby  degree_centrality  \
92405              0.581977       2.309477           0.016807   
90957              1.292540      10.763777           0.016807   
112128             0.370873      12.479752           0.033613   
90608              0.776182       6.297489           0.033613   
106192             0.658035       3.807573           0.016807   
91871              0.693391       2.058774           0.016807   
93825              0.451637       2.594828           0.016807   
92504              0.468378       2.365532           0.016807   
105695             0.745596       3.720593           0.016807   
90432              0.827889       6.370404           0.033613   
92299              0.490926       2.278805           0.016807   
92442              0.686789       2.664024           0.016807   
93895              0.490378       2.630876           0.016807   
105699             0.745596       3.720593           0.016807   
88081              0.860056       7.263401           0.016807   
106057             0.584731       3.882019           0.016807   
90521              0.947205       6.663943           0.033613   
92340              0.451387       2.128424           0.016807   
91694              0.481551       2.262574           0.016807   
98379              0.899849       8.828235           0.016807   
92226              0.586629       2.932814           0.016807   
105468             0.701852       3.763948           0.016807   
92533              0.656363       1.982722           0.016807   
105966             0.745596       3.720593           0.016807   
92066              0.538563       2.345844           0.016807   

        eigenvector_centrality                 flat_model_type  \
92405                 0.047782                3 ROOM, Standard   
90957                 0.000217            EXECUTIVE, Apartment   
112128                0.000229        

\# TODO: \<Enter your answer here\>